crates/bevy_ui/src/stack.rs

```diff
@@ -15,138 +15,78 @@ pub struct UiStack {
-        // Reserve space for all children. In practice, some may not get pushed since
-        // nodes with `ZIndex::Global` are pushed to the global (root) context.
-        new_context.entries.reserve_exact(children.len());
-
-        for entity in children {
-            insert_context_hierarchy(
-                cache,
-                zindex_query,
-                children_query,
-                *entity,
-                global_context,
-                Some(&mut new_context),
-                total_entry_count,
-            );
+    radsort::sort_by_key(&mut root_nodes, |(_, z)| *z);
+    traversal_stack.extend(root_nodes.into_iter().map(|(id, _)| id).rev());
+
+    while let Some(entity) = traversal_stack.pop() {
+        uinodes.push(entity);
+
+        if let Ok(children) = children_query.get(entity) {
+            let mut z_children: Vec<(Entity, i32)> = children
+                .iter()
+                .filter_map(|entity| {
+                    zindex_query
+                        .get(*entity)
+                        .ok()
+                        .map(|zindex| (*entity, zindex.map(|zindex| zindex.0).unwrap_or(0)))
+                })
+                .collect();
+            radsort::sort_by_key(&mut z_children, |k| k.1);
+
+            // reverse order because we're walking the tree depth first
+            for (child_id, _) in z_children.into_iter().rev() {
+                traversal_stack.push(child_id);
+            }
         }
     }
 
-    // The node will be added either to global/parent based on its z-index type: global/local.
-    let z_index = zindex_query.get(entity).unwrap_or(&ZIndex::Local(0));
-    let (entity_context, z_index) = match z_index {
-        ZIndex::Local(value) => (parent_context.unwrap_or(global_context), *value),
-        ZIndex::Global(value) => (global_context, *value),
-    };
-
-    *total_entry_count += 1;
-    entity_context.entries.push(StackingContextEntry {
-        z_index,
-        entity,
-        stack: new_context,
-    });
-}
-
-/// Flatten `StackingContext` (z-index based UI node tree) into back-to-front entities list
-fn fill_stack_recursively(
-    cache: &mut StackingContextCache,
-    result: &mut Vec<Entity>,
-    stack: &mut StackingContext,
-) {
-    // Sort entries by ascending z_index, while ensuring that siblings
-    // with the same local z_index will keep their ordering. This results
-    // in `back-to-front` ordering, low z_index = back; high z_index = front.
-    stack.entries.sort_by_key(|e| e.z_index);
-
-    for mut entry in stack.entries.drain(..) {
-        // Parent node renders before/behind child nodes
-        result.push(entry.entity);
-        fill_stack_recursively(cache, result, &mut entry.stack);
-        cache.push(entry.stack);
+    for (i, entity) in ui_stack.uinodes.iter().enumerate() {
+        if let Ok(mut node) = update_query.get_mut(*entity) {
+            node.bypass_change_detection().stack_index = i as u32;
+        }
     }
```

Comment:
The `traversal_stack` is never inserted to `ui_stack`. Am I missing something here? Also how are children supposed to be inserted relative to their parents within the global stack?

Reply:
I'm puzzled how this passes CI, it appears to have serious bugs.

Reply:
`traversal_stack` is used to navigate the tree, not insert the nodes.

Reply:
Abandoned the iterative version using a stack anyway, and went back to walking the tree recursively copying your stack of cached vecs approach from main. I don't understand why the iterative version is so much less efficient though. It just has one buffer for storing and sorting the children it reuses, which seems so much better.